FreeRTOS_Sockets.c

```diff
@@ -3647,155 +3935,214 @@ void vSocketWakeUpUser( FreeRTOS_Socket_t * pxSocket )
             {
-                xByteCount = 0;
+                /* Return -ENOTCONN, unless there was a malloc failure. */
+                xByteCount = -pdFREERTOS_ERRNO_ENOTCONN;
+
+                if( pxSocket->u.xTCP.bits.bMallocError != pdFALSE_UNSIGNED )
+                {
+                    /* The no-memory error has priority above the non-connected error.
+                     * Both are fatal and will lead to closing the socket. */
+                    xByteCount = -pdFREERTOS_ERRNO_ENOMEM;
+                }
+
+                break;
             }
 
-            while( xByteCount == 0 )
+            if( xTimed == pdFALSE )
             {
-                switch( ipNUMERIC_CAST( eIPTCPState_t, pxSocket->u.xTCP.ucTCPState ) )
-                {
-                    case eCLOSED:
-                    case eCLOSE_WAIT: /* (server + client) waiting for a connection termination request from the local user. */
-                    case eCLOSING:    /* (server + client) waiting for a connection termination request acknowledgement from the remote TCP. */
+                /* Only in the first round, check for non-blocking. */
+                xRemainingTime = pxSocket->xReceiveBlockTime;
 
-                        if( pxSocket->u.xTCP.bits.bMallocError != pdFALSE_UNSIGNED )
-                        {
-                            /* The no-memory error has priority above the non-connected error.
-                             * Both are fatal and will lead to closing the socket. */
-                            xByteCount = -pdFREERTOS_ERRNO_ENOMEM;
-                        }
-                        else
+                if( xRemainingTime == ( TickType_t ) 0 )
+                {
+                    #if ( ipconfigSUPPORT_SIGNALS != 0 )
                         {
-                            xByteCount = -pdFREERTOS_ERRNO_ENOTCONN;
+                            /* Just check for the interrupt flag. */
+                            xEventBits = xEventGroupWaitBits( pxSocket->xEventGroup, ( EventBits_t ) eSOCKET_INTR,
+                                                              pdTRUE /*xClearOnExit*/, pdFALSE /*xWaitAllBits*/, socketDONT_BLOCK );
                         }
-
-                        break;
-
-                    case eTCP_LISTEN:
-                    case eCONNECT_SYN:
-                    case eSYN_FIRST:
-                    case eSYN_RECEIVED:
-                    case eESTABLISHED:
-                    case eFIN_WAIT_1:
-                    case eFIN_WAIT_2:
-                    case eLAST_ACK:
-                    case eTIME_WAIT:
-                    default:
-                        /* Nothing. */
-                        break;
+                    #endif /* ipconfigSUPPORT_SIGNALS */
+                    break;
                 }
 
-                if( xByteCount < 0 )
+                if( ( ( uint32_t ) xFlags & ( uint32_t ) FREERTOS_MSG_DONTWAIT ) != 0U )
```

Comment:
Perhaps xFlags should be a uint32_t.

Reply:
In the original BSD declarations, `flags` were declared as `int`.
As you know, MISRA doesn't like to do bit-testing in a signed integer.
So therefore the use of a cast.